File: android/lca/src/main/java/com/dreampany/lca/data/source/remote/GraphRemoteDataSource.java
package com.dreampany.lca.data.source.remote;

import com.dreampany.lca.data.misc.GraphMapper;
import com.dreampany.lca.data.model.Graph;
import com.dreampany.lca.data.source.api.GraphDataSource;
import com.dreampany.lca.misc.CmcGraphAnnote;
import com.dreampany.network.manager.NetworkManager;

import java.util.List;

import javax.inject.Singleton;

import io.reactivex.Maybe;

/**
 * Created by Hawladar Roman on 30/5/18.
 * Dreampany Ltd
 * dreampanymail@gmail.com
 */
@Singleton
public class GraphRemoteDataSource implements GraphDataSource {

    private final NetworkManager network;
    private final GraphMapper mapper;
    private final CoinMarketCapGraphService service;

    public GraphRemoteDataSource(NetworkManager network,
                                 GraphMapper mapper,
                                 @CmcGraphAnnote CoinMarketCapGraphService service) {
        this.network = network;
        this.mapper = mapper;
        this.service = service;
    }

    @Override
    public boolean isEmpty() {
        return false;
    }

    @Override
    public Maybe<Boolean> isEmptyRx() {
        return Maybe.fromCallable(this::isEmpty);
    }

    @Override
    public int getCount() {
        return 0;
    }

    @Override
    public Maybe<Integer> getCountRx() {
        return null;
    }

    @Override
    public boolean isExists(Graph graph) {
        return false;
    }

    @Override
    public Maybe<Boolean> isExistsRx(Graph graph) {
        return null;
    }

    @Override
    public long putItem(Graph graph) {
        return 0;
    }

    @Override
    public Maybe<Long> putItemRx(Graph graph) {
        return null;
    }

    @Override
    public List<Long> putItems(List<? extends Graph> graphs) {
        return null;
    }

    @Override
    public Maybe<List<Long>> putItemsRx(List<? extends Graph> graphs) {
        return null;
    }

    @Override
    public int delete(Graph graph) {
        return 0;
    }

    @Override
    public Maybe<Integer> deleteRx(Graph graph) {
        return null;
    }

    @Override
    public List<Long> delete(List<? extends Graph> graphs) {
        return null;
    }

    @Override
    public Maybe<List<Long>> deleteRx(List<? extends Graph> graphs) {
        return null;
    }

    @Override
    public Graph getItem(String id) {
        return null;
    }

    @Override
    public Maybe<Graph> getItemRx(String id) {
        return null;
    }

    @Override
    public List<Graph> getItems() {
        return null;
    }

    @Override
    public Maybe<List<Graph>> getItemsRx() {
        return null;
    }

    @Override
    public List<Graph> getItems(int limit) {
        return null;
    }

    @Override
    public Maybe<List<Graph>> getItemsRx(int limit) {
        return null;
    }

    @Override
    public Graph getItem(String slug, long startTime, long endTime) {
        return null;
    }

    @Override
    public Maybe<Graph> getItemRx(String slug, long startTime, long endTime) {
        return service
                .getGraphRx(slug, startTime, endTime)
                .map(cmcGraph -> {
                    cmcGraph.setStartTime(startTime);
                    cmcGraph.setEndTime(endTime);
                    return mapper.toGraph(cmcGraph);
                });
    }
}

File: android/tools/src/main/java/com/dreampany/tools/data/source/crypto/remote/GraphRemoteDataSource.kt
package com.dreampany.tools.data.source.crypto.remote

import android.content.Context
import com.dreampany.framework.misc.func.Keys
import com.dreampany.framework.misc.func.Parser
import com.dreampany.framework.misc.func.SmartError
import com.dreampany.network.manager.NetworkManager
import com.dreampany.tools.api.crypto.remote.service.CoinMarketCapGraphService
import com.dreampany.tools.data.model.crypto.Graph
import com.dreampany.tools.data.source.crypto.api.GraphDataSource
import com.dreampany.tools.data.source.crypto.mapper.GraphMapper
import java.net.UnknownHostException

/**
 * Created by roman on 10/11/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
class GraphRemoteDataSource
constructor(
private val context: Context,
private val network: NetworkManager,
private val parser: Parser,
private val keys: Keys,
private val mapper : GraphMapper,
private val service: CoinMarketCapGraphService
) : GraphDataSource {

    @Throws
    override suspend fun read(slug: String, startTime: Long, endTime: Long): Graph? {
        try {
             val response= service.read(slug, startTime, endTime).execute()
            if (response.isSuccessful) {
                val data = response.body() ?: return null
                return mapper.read(data)
            } else {
                throw SmartError(

                )
                /*val error = parser.parseError(response, NewsResponse::class)
                throw SmartError(
                    message = error?.status?.errorMessage,
                    code = error?.status?.errorCode
                )*/
            }
        } catch (error: Throwable) {
            if (error is SmartError) throw error
            if (error is UnknownHostException) throw SmartError(
                message = error.message,
                error = error
            )
        }
        throw SmartError()
    }
}